examples/hotrod/pkg/log/spanlogger.go

```diff
@@ -57,99 +58,164 @@ func (sl spanLogger) With(fields ...zapcore.Field) Logger {
 	return spanLogger{logger: sl.logger.With(fields...), span: sl.span, spanFields: sl.spanFields}
 }
 
-func (sl spanLogger) logToSpan(level string, msg string, fields ...zapcore.Field) {
-	// TODO rather than always converting the fields, we could wrap them into a lazy logger
-	fa := fieldAdapter(make([]log.Field, 0, 2+len(fields)))
-	fa = append(fa, log.String("event", msg))
-	fa = append(fa, log.String("level", level))
+func (sl spanLogger) logToSpan(msg string, fields ...zapcore.Field) {
+	sl.span.AddEvent(
+		msg,
+		trace.WithAttributes(otLogFieldsToOTelAttrs(fields)...),
+	)
+}
+
+func otLogFieldsToOTelAttrs(fields []zapcore.Field) []attribute.KeyValue {
+	encoder := &bridgeFieldEncoder{}
 	for _, field := range fields {
-		field.AddTo(&fa)
+		field.AddTo(encoder)
 	}
-	sl.span.LogFields(fa...)
+	return encoder.pairs
+}
+
+type bridgeFieldEncoder struct {
+	pairs []attribute.KeyValue
+}
+
+func (e *bridgeFieldEncoder) AddArray(key string, marshaler zapcore.ArrayMarshaler) error {
+	e.pairs = append(e.pairs, otTagToOTelAttr(key, marshaler))
+	return nil
+}
+
+func (e *bridgeFieldEncoder) AddObject(key string, marshaler zapcore.ObjectMarshaler) error {
+	e.pairs = append(e.pairs, otTagToOTelAttr(key, marshaler))
+	return nil
 }
 
-type fieldAdapter []log.Field
+func (e *bridgeFieldEncoder) AddBinary(key string, value []byte) {
+	e.pairs = append(e.pairs, otTagToOTelAttr(key, value))
```

Comment:
Since we know the type here, why do we call into method that erases the type and has to match on it at runtime? We could call the right otel attribute constructor right here.

Reply:
Just out of convenience, particularly for the "unsupported" types which fall back to the `default` case; although I like the fact that there's less indirection, so I've made direct calls in: [b1d21a1](https://github.com/jaegertracing/jaeger/pull/4556/commits/b1d21a13dee7d7083ce6873a0f0c1bd9d4557bba).